File: Swift/backspace-string-compare.swift
class Solution {
    func backspaceCompare(_ S: String, _ T: String) -> Bool {
        return getResultString(s: S) == getResultString(s: T)
    }

    private func getResultString(s: String) -> String {
        var chars = Array(s)
        var res = Array<Character>()
        var skip = 0

        for i in Array(0..<chars.count).reversed() {
            if chars[i] == "#" {
                skip += 1
            } else if skip > 0 {
                skip -= 1
            } else {
                res.insert(chars[i], at: 0)
            }
        }

        return String(res)
    }
}

File: Python/backspace-string-compare.py
import itertools

class Solution:
    def backspaceCompare(self, S, T):
        """
        :type S: str
        :type T: str
        :rtype: bool
        """
        return all(x == y for x, y in
                   itertools.zip_longest(self.getResultString(S), self.getResultString(T)))

    def getResultString(self, S):
        skip = 0
        for i in reversed(range(len(S))):
            if S[i] == '#':
                skip += 1
            elif skip:
                skip -= 1
            else:
                yield S[i]
